src/inc/til/small_vector.h

```diff
@@ -0,0 +1,864 @@
+            tmp += off;
+            return tmp;
+        }
+
+        constexpr small_vector_iterator& operator-=(const difference_type off) noexcept
+        {
+            base::operator-=(off);
+            return *this;
+        }
+
+        using base::operator-;
+
+        [[nodiscard]] constexpr small_vector_iterator operator-(const difference_type off) const noexcept
+        {
+            small_vector_iterator tmp = *this;
+            tmp -= off;
+            return tmp;
+        }
+
+        [[nodiscard]] constexpr reference operator[](const difference_type off) const noexcept
+        {
+            return const_cast<reference>(base::operator[](off));
+        }
+
+        using _Prevent_inheriting_unwrap = small_vector_iterator;
+
+        [[nodiscard]] constexpr pointer _Unwrapped() const noexcept
+        {
+#pragma warning(suppress : 26492) // Don't use const_cast to cast away const or volatile (type.3).
+            return const_cast<pointer>(base::_Unwrapped());
+        }
+    };
+
+    template<typename T, size_t N>
+    class small_vector
+    {
+    public:
+        static_assert(N != 0, "A small_vector without a small buffer isn't very useful");
+        static_assert(std::is_nothrow_move_assignable_v<T>, "_generic_insert doesn't guard against exceptions");
+        static_assert(std::is_nothrow_move_constructible_v<T>, "_grow/_generic_insert don't guard against exceptions");
+
+        using value_type = T;
+        using allocator_type = std::allocator<T>;
+        using pointer = T*;
+        using const_pointer = const T*;
+        using reference = T&;
+        using const_reference = const T&;
+        using size_type = size_t;
+        using difference_type = ptrdiff_t;
+
+        using iterator = small_vector_iterator<T>;
+        using const_iterator = small_vector_const_iterator<T>;
+        using reverse_iterator = std::reverse_iterator<iterator>;
+        using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+        small_vector() noexcept :
+            _data{ &_buffer[0] }
+        {
+        }
+
+        explicit small_vector(size_type count, const T& value = T{}) :
+            small_vector{}
+        {
+            insert(end(), count, value);
+        }
+
+        template<typename InputIt>
+        small_vector(InputIt first, InputIt last) :
+            small_vector{}
+        {
+            insert(end(), first, last);
+        }
+
+        small_vector(std::initializer_list<T> list) :
+            small_vector{}
+        {
+            insert(end(), list.begin(), list.end());
+        }
+
+        // NOTE: If an exception is thrown while copying, the vector is left empty.
+        small_vector(const small_vector& other) :
+            small_vector{}
+        {
+            operator=(other);
+        }
+
+        // NOTE: If an exception is thrown while copying, the vector is left empty.
+        small_vector& operator=(const small_vector& other)
+        {
+            clear();

```

Comment:
`clear()` only destroys the data, but doesn't free the allocated memory.
The copy constructor wants exactly that: It wants to reuse the existing memory.
But the move constructor needs to free its own memory, since it'll take ownership of the memory of the `other` vector.

Reply:
Ah, I was thinking about calling the destructors of the objects inside when I wrote this comment not necessarily about using the memory space.